File: RangeCalc.go/RangeCalc.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

const ProgVer = "2.0.1 - Full Triangulation"

// stdin is shared across all prompts. bufio.Reader buffers ahead of what it's
// asked for, so creating a fresh one per prompt (as this used to do) silently
// discards any input already buffered past the first line — piped input past
// the first prompt would vanish.
var stdin = bufio.NewReader(os.Stdin)

type TriangulationResult struct {
	RangeFromRef1 float64
	RangeFromRef2 float64
	AngleAtTarget float64
}

func main() {
	b1Flag := flag.Float64("b1", 0, "Bearing from Ref One to Target, degrees (1-360)")
	b2Flag := flag.Float64("b2", 0, "Bearing from Ref Two to Target, degrees (1-360)")
	dFlag := flag.Float64("d", 0, "Baseline distance between Ref One and Ref Two, meters")
	bpFlag := flag.Float64("bp", 0, "Baseline bearing, Ref One to Ref Two, degrees (1-360) — selects Full Triangulation")
	oFlag := flag.String("o", "", "Write a formatted report to this file instead of printing bare results to stdout")
	flag.Parse()

	set := map[string]bool{}
	flag.Visit(func(f *flag.Flag) { set[f.Name] = true })

	if len(set) > 0 {
		runNonInteractive(*b1Flag, *b2Flag, *dFlag, *bpFlag, *oFlag, set["b1"], set["b2"], set["d"], set["bp"])
		return
	}

	printHeader()

	bearing1 := getBearing("Bearing from Ref One to Target")
	bearing2 := getBearing("Bearing from Ref Two to Target")
	baseline := getPositiveDistance("Baseline Distance (Ref1 to Ref2 in meters)")

	fmt.Println("\nChoose calculation method:")
	fmt.Println("1. Simple (fast, assumes perpendicular baseline)")
	fmt.Println("2. Full Triangulation (more accurate with Baseline Bearing)")
	fmt.Print("Enter choice (1 or 2): ")
	choice := readInput()

	var result TriangulationResult
	if choice == "2" {
		baselineBearing := getBearing("Baseline Bearing (direction from Ref1 to Ref2)")
		result = calculateFullTriangulation(bearing1, bearing2, baseline, baselineBearing)
	} else {
		result = calculateSimple(bearing1, bearing2, baseline)
	}

	fmt.Println("\n=== RESULTS ===")
	fmt.Printf("Range from Ref One : %.2f meters\n", result.RangeFromRef1)
	fmt.Printf("Range from Ref Two : %.2f meters\n", result.RangeFromRef2)
	fmt.Printf("Angle at Target    : %.1f°\n", result.AngleAtTarget)
}

// runNonInteractive handles the -b1/-b2/-d/-bp/-o flag-driven path, for
// scripting and piping. -b1, -b2, and -d are required together; -bp is
// optional and selects Full Triangulation in place of Simple mode; -o is
// optional and redirects a formatted report to a file instead of printing
// bare key=value results to stdout.
func runNonInteractive(b1, b2, baseline, bp float64, outFile string, b1Set, b2Set, dSet, bpSet bool) {
	if !b1Set || !b2Set || !dSet {
		fmt.Fprintln(os.Stderr, "Error: -b1, -b2, and -d are all required together.")
		flag.Usage()
		os.Exit(2)
	}
	if b1 < 1 || b1 > 360 {
		fmt.Fprintln(os.Stderr, "Error: -b1 must be between 1 and 360 degrees.")
		os.Exit(2)
	}
	if b2 < 1 || b2 > 360 {
		fmt.Fprintln(os.Stderr, "Error: -b2 must be between 1 and 360 degrees.")
		os.Exit(2)
	}
	if baseline <= 0 {
		fmt.Fprintln(os.Stderr, "Error: -d must be greater than zero.")
		os.Exit(2)
	}
	if bpSet && (bp < 1 || bp > 360) {
		fmt.Fprintln(os.Stderr, "Error: -bp must be between 1 and 360 degrees.")
		os.Exit(2)
	}

	method := "Simple"
	var result TriangulationResult
	if bpSet {
		method = "Full Triangulation"
		result = calculateFullTriangulation(b1, b2, baseline, bp)
	} else {
		result = calculateSimple(b1, b2, baseline)
	}

	if outFile == "" {
		fmt.Printf("RangeFromRef1=%.2f\n", result.RangeFromRef1)
		fmt.Printf("RangeFromRef2=%.2f\n", result.RangeFromRef2)
		fmt.Printf("AngleAtTarget=%.1f\n", result.AngleAtTarget)
		return
	}

	report := buildReport(b1, b2, baseline, bp, bpSet, method, result)
	if err := os.WriteFile(outFile, []byte(report), 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing report to %s: %v\n", outFile, err)
		os.Exit(1)
	}
	fmt.Printf("Report written to %s\n", outFile)
}

const reportWidth = 72

func boxRule(width int) string {
	return "+" + strings.Repeat("-", width-2) + "+\n"
}

func boxCentered(text string, width int) string {
	pad := width - 2 - len(text)
	if pad < 0 {
		pad = 0
	}
	left := pad / 2
	right := pad - left
	return "|" + strings.Repeat(" ", left) + text + strings.Repeat(" ", right) + "|\n"
}

// buildReport formats a full result report for -o file output: an ASCII-box
// heading followed by the inputs used and the calculated results.
func buildReport(b1, b2, baseline, bp float64, bpSet bool, method string, result TriangulationResult) string {
	var sb strings.Builder
	sb.WriteString(boxRule(reportWidth))
	sb.WriteString(boxCentered(fmt.Sprintf("RangeCalc v%s", ProgVer), reportWidth))
	sb.WriteString(boxRule(reportWidth))
	sb.WriteString("\n")

	sb.WriteString(fmt.Sprintf("Method            : %s\n", method))
	sb.WriteString(fmt.Sprintf("Bearing 1         : %.1f°\n", b1))
	sb.WriteString(fmt.Sprintf("Bearing 2         : %.1f°\n", b2))
	sb.WriteString(fmt.Sprintf("Baseline Distance : %.1f m\n", baseline))
	if bpSet {
		sb.WriteString(fmt.Sprintf("Baseline Bearing  : %.1f°\n", bp))
	}
	sb.WriteString("\n")

	sb.WriteString(boxRule(reportWidth))
	sb.WriteString(boxCentered("RESULTS", reportWidth))
	sb.WriteString(boxRule(reportWidth))
	sb.WriteString(fmt.Sprintf("Range from Ref One : %.2f meters\n", result.RangeFromRef1))
	sb.WriteString(fmt.Sprintf("Range from Ref Two : %.2f meters\n", result.RangeFromRef2))
	sb.WriteString(fmt.Sprintf("Angle at Target    : %.1f°\n", result.AngleAtTarget))

	return sb.String()
}

func printHeader() {
	fmt.Printf("+------------------------------[%s]----------------------------+\n", ProgVer)
	fmt.Println("|        Range Calculator - Full Triangulation Edition           |")
	fmt.Println("| View the code on GitHub [https://github.com/JohnDovey/RangeCalc] |")
	fmt.Println("+--------------------------------------------------------------------+\n")
	fmt.Println("Enter the values as prompted (0 to exit)\n")
}

func readInput() string {
	input, err := stdin.ReadString('\n')
	trimmed := strings.TrimSpace(input)
	if err != nil && trimmed == "" {
		fmt.Println("\nNo more input available. Exiting.")
		os.Exit(0)
	}
	return trimmed
}

func getBearing(prompt string) float64 {
	for {
		fmt.Printf("%s: ", prompt)
		input := readInput()

		if input == "0" {
			fmt.Println("Operation cancelled.")
			os.Exit(0)
		}

		value, err := strconv.ParseFloat(input, 64)
		if err != nil {
			fmt.Println("Invalid input. Please enter a number.")
			continue
		}

		if value < 1 || value > 360 {
			fmt.Println("Error: Bearing must be between 1 and 360 degrees.")
			continue
		}

		fmt.Printf("\t(%.0f°)\n", value)
		return value
	}
}

func getPositiveDistance(prompt string) float64 {
	for {
		fmt.Printf("%s: ", prompt)
		input := readInput()

		if input == "0" {
			fmt.Println("Operation cancelled.")
			os.Exit(0)
		}

		value, err := strconv.ParseFloat(input, 64)
		if err != nil || value < 1 {
			fmt.Println("Error: Please enter a positive number greater than zero.")
			continue
		}

		fmt.Printf("\t(%.0f meters)\n", value)
		return value
	}
}

func calculateSimple(b1, b2, baseline float64) TriangulationResult {
	diff := math.Abs(b1 - b2)
	if diff > 180 {
		diff = 360 - diff
	}
	if diff == 0 {
		fmt.Println("Error: Bearings are identical.")
		os.Exit(1)
	}

	rad := (90.0 - diff) * math.Pi / 180.0
	rangeVal := baseline * math.Tan(rad)

	return TriangulationResult{
		RangeFromRef1: rangeVal,
		RangeFromRef2: rangeVal,
		AngleAtTarget: diff,
	}
}

func calculateFullTriangulation(b1, b2, baseline, baselineBearing float64) TriangulationResult {
	angleAtTarget := math.Abs(b1 - b2)
	if angleAtTarget > 180 {
		angleAtTarget = 360 - angleAtTarget
	}

	if angleAtTarget == 0 {
		fmt.Println("Error: Bearings are identical.")
		os.Exit(1)
	}

	angleAtRef1 := math.Abs(baselineBearing - b1)
	if angleAtRef1 > 180 {
		angleAtRef1 = 360 - angleAtRef1
	}

	angleAtRef2 := 180.0 - angleAtTarget - angleAtRef1

	if angleAtRef2 <= 0 {
		fmt.Println("Error: Invalid geometry (target position not possible with these bearings).")
		os.Exit(1)
	}

	// Law of Sines
	sinTarget := math.Sin(angleAtTarget * math.Pi / 180)
	rangeFromRef1 := baseline * math.Sin(angleAtRef2*math.Pi/180) / sinTarget
	rangeFromRef2 := baseline * math.Sin(angleAtRef1*math.Pi/180) / sinTarget

	return TriangulationResult{
		RangeFromRef1: rangeFromRef1,
		RangeFromRef2: rangeFromRef2,
		AngleAtTarget: angleAtTarget,
	}
}

File: RangeCalcWin/rangecalc.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <cmath>
#include <limits>
#include <stdexcept>

const std::string ProgVer = "2.0.0";

// Portable pi (MSVC does not always define M_PI)
const double PI = 3.14159265358979323846;

void PrintHeader()
{
    std::cout << "+------------------------------[" << std::setw(6) << ProgVer << "]----------------------------+\n";
    std::cout << "|        Range Calculator - by John Dovey <dovey.john@gmail.com>     |\n";
    std::cout << "| View the code on GitHub [https://github.com/JohnDovey/RangeCalc]   |\n";
    std::cout << "+--------------------------------------------------------------------+\n\n";
}

// Smallest angle between two compass bearings, in (0, 180]
double CircularAngleDiff(double a, double b)
{
    double diff = std::abs(a - b);
    if (diff > 180.0)
        diff = 360.0 - diff;
    return diff;
}

double DegToRad(double deg)
{
    return deg * PI / 180.0;
}

int GetBearing(const std::string& prompt)
{
    while (true)
    {
        std::cout << prompt << ": ";
        std::string input;
        std::getline(std::cin, input);

        try
        {
            int value = std::stoi(input);

            if (value == 0)
                throw std::runtime_error("Operation cancelled.");

            if (value < 1 || value > 360)
            {
                std::cout << "Error: Bearing must be between 1 and 360 degrees.\n\n";
                continue;
            }

            std::cout << "\t(" << value << " deg)\n";
            return value;
        }
        catch (const std::runtime_error&)
        {
            throw; // cancellation
        }
        catch (...)
        {
            std::cout << "Invalid input. Please enter a number.\n\n";
        }
    }
}

int GetPositiveDistance(const std::string& prompt)
{
    while (true)
    {
        std::cout << prompt << ": ";
        std::string input;
        std::getline(std::cin, input);

        try
        {
            int value = std::stoi(input);

            if (value == 0)
                throw std::runtime_error("Operation cancelled.");

            if (value < 1)
            {
                std::cout << "Error: Distance must be greater than zero.\n\n";
                continue;
            }

            std::cout << "\t(" << value << " meters between reference points)\n";
            return value;
        }
        catch (const std::runtime_error&)
        {
            throw;
        }
        catch (...)
        {
            std::cout << "Invalid input. Please enter a positive number.\n\n";
        }
    }
}

struct TriangulationResult
{
    double rangeFromRef1;
    double rangeFromRef2;
    double angleAtTarget;
};

// Simple mode: range = baseline * cot(theta) = baseline * tan(90 - theta)
// where theta is the circular bearing difference (parallax angle).
// Assumes a symmetric / "perpendicular enough" geometry; approximate only.
// Valid for 0 < theta < 90 degrees.
TriangulationResult CalculateSimple(double b1, double b2, double baseline)
{
    double theta = CircularAngleDiff(b1, b2);

    if (theta == 0.0)
        throw std::runtime_error("Bearings are identical - cannot calculate range.");

    if (theta >= 90.0)
        throw std::runtime_error(
            "Angle difference too large for simple mode (need < 90 deg). "
            "Try Full Triangulation with a baseline bearing.");

    // cot(theta) = tan(90 - theta); convert degrees -> radians for std::tan
    double range = baseline * std::tan(DegToRad(90.0 - theta));

    return { range, range, theta };
}

// Full triangulation via law of sines.
// Requires baseline bearing (direction from Ref1 to Ref2) so the triangle
// angles at each reference point can be formed from absolute compass bearings.
TriangulationResult CalculateFull(double b1, double b2, double baseline, double baselineBearing)
{
    double angleAtTarget = CircularAngleDiff(b1, b2);

    if (angleAtTarget == 0.0)
        throw std::runtime_error("Bearings are identical - cannot calculate range.");

    // Interior angle at Ref1: between baseline (Ref1->Ref2) and LOB (Ref1->Target)
    double angleAtRef1 = CircularAngleDiff(baselineBearing, b1);

    // Third angle of the plane triangle
    double angleAtRef2 = 180.0 - angleAtTarget - angleAtRef1;

    if (angleAtRef1 <= 0.0 || angleAtRef2 <= 0.0)
        throw std::runtime_error(
            "Invalid geometry - target position not possible with these bearings.");

    // Law of sines: a/sin(A) = b/sin(B) = c/sin(C)
    //   range_from_ref1 / sin(angleAtRef2) = baseline / sin(angleAtTarget)
    double sinTarget = std::sin(DegToRad(angleAtTarget));
    if (std::abs(sinTarget) < 1e-12)
        throw std::runtime_error("Degenerate triangle (angle at target near 0 or 180).");

    double rangeFromRef1 = baseline * std::sin(DegToRad(angleAtRef2)) / sinTarget;
    double rangeFromRef2 = baseline * std::sin(DegToRad(angleAtRef1)) / sinTarget;

    if (rangeFromRef1 <= 0.0 || rangeFromRef2 <= 0.0)
        throw std::runtime_error("Invalid geometry - computed range is not positive.");

    return { rangeFromRef1, rangeFromRef2, angleAtTarget };
}

int main()
{
    PrintHeader();
    std::cout << "Enter the values as prompted (0 to exit)\n\n";

    try
    {
        int bearing1 = GetBearing("Bearing from Ref One to Target");
        int bearing2 = GetBearing("Bearing from Ref Two to Target");
        int refDistance = GetPositiveDistance("Distance between Ref One and Ref Two");

        std::cout << "\nChoose calculation method:\n";
        std::cout << "1. Simple (fast; assumes favourable geometry)\n";
        std::cout << "2. Full Triangulation (accurate; needs baseline bearing)\n";
        std::cout << "Enter choice (1 or 2): ";

        std::string choice;
        std::getline(std::cin, choice);

        TriangulationResult result;
        if (choice == "2")
        {
            int baselineBearing = GetBearing("Baseline Bearing (direction from Ref1 to Ref2)");
            result = CalculateFull(bearing1, bearing2, refDistance, baselineBearing);
        }
        else
        {
            result = CalculateSimple(bearing1, bearing2, refDistance);
        }

        std::cout << "\n";
        std::cout << std::fixed << std::setprecision(2);
        std::cout << "=== RESULTS ===\n";
        std::cout << "Range from Ref One : " << result.rangeFromRef1 << " meters\n";
        std::cout << "Range from Ref Two : " << result.rangeFromRef2 << " meters\n";
        std::cout << std::setprecision(1);
        std::cout << "Angle at Target    : " << result.angleAtTarget << " deg\n\n";
    }
    catch (const std::exception& e)
    {
        std::cout << "\nError: " << e.what() << std::endl;
    }

    std::cout << "Press Enter to exit...";
    std::cin.get();
    return 0;
}
